core/src/main/kotlin/fr/sncf/osrd/standalone_sim/etcsContextBuilder.kt

```diff
@@ -0,0 +1,121 @@
+            if (infra.getTrackNodeConfigs(element).size <= 1U) continue
+            res.add(Offset(position.distance + currentZonePathStartOffset))
+        }
+        currentZonePathStartOffset += infra.getZonePathLength(zonePath).distance
+    }
+
+    findLastDangerPoint(infra, chunkPath)?.let { res.add(it) }
+    return res.sorted()
+}
+
+/**
+ * Find the last danger point, which may extend beyond the end of the path. Null if tracks are
+ * circular with no switch nor buffer stop.
+ */
+fun findLastDangerPoint(infra: RawInfra, chunkPath: ChunkPath): Offset<Path>? {
+    // Find the offset of the last chunk on the path
+    val lastChunk = chunkPath.chunks.last()
+    var lastChunkEndOffset = chunkPath.beginOffset * -1.0
+    for (chunk in chunkPath.chunks) {
+        lastChunkEndOffset += infra.getTrackChunkLength(chunk.value).distance
+    }
+    val lastTrack = infra.getTrackFromChunk(lastChunk.value)
+    val endOfLastTrackPathOffset =
+        getEndOfLastTrackPathOffset(infra, lastTrack, lastChunk, lastChunkEndOffset)
+
+    // Iterate on the tracks until there either a switch or buffer stop
+    var currentTrackEndOffset = endOfLastTrackPathOffset
+    var track = DirStaticIdx(lastTrack, lastChunk.direction)
+    while (true) {
+        val nextTracks = infra.getNextTrackSections(track)
+        val endAtDangerPoint = nextTracks.size != 1
+        if (endAtDangerPoint) {
+            return currentTrackEndOffset
+        }
+        track = nextTracks.single()
+        currentTrackEndOffset += infra.getTrackSectionLength(track.value).distance
+        if (track.value == lastTrack) return null // Circular tracks
+    }
+}
+
+/** Figure out where the end of the last track is located, as a path offset. */
+fun getEndOfLastTrackPathOffset(
+    infra: RawInfra,
+    lastTrack: TrackSectionId,
+    lastChunk: DirStaticIdx<TrackChunk>,
+    lastChunkEndOffset: Offset<Path>
+): Offset<Path> {
+    val lastTrackLength = infra.getTrackSectionLength(lastTrack)
+    val lastChunkLength = infra.getTrackChunkLength(lastChunk.value)
+
+    // As an offset on the undirected last track, where the start of the (undirected) last chunk is
+    // located
+    val lastUndirectedChunkStartOffsetOnTrack = infra.getTrackChunkOffset(lastChunk.value)
+
+    // As an offset on the undirected last track, location of the start of the directed last chunk
+    val lastDirectedChunkStartOffsetOnTrack =
+        if (lastChunk.direction == Direction.INCREASING) lastUndirectedChunkStartOffsetOnTrack
+        else lastUndirectedChunkStartOffsetOnTrack + lastChunkLength.distance
+
+    val distanceFromChunkEndToTrackEnd =
+        if (lastChunk.direction == Direction.INCREASING)
+            lastTrackLength.distance -
+                (lastDirectedChunkStartOffsetOnTrack.distance + lastChunkLength.distance)
+        else lastDirectedChunkStartOffsetOnTrack.distance - lastChunkLength.distance
```

Comment:
You may also save accessor calls in `DECREASING` case, but you can at least simplify:
```suggestion
    val distanceFromChunkEndToTrackEnd =
        if (lastChunk.direction == Direction.INCREASING)
            lastTrackLength.distance -
                (lastUndirectedChunkStartOffsetOnTrack.distance + lastChunkLength.distance)
        else lastUndirectedChunkStartOffsetOnTrack.distance
```